opensearch/sycamore-opensearch.sh

```diff
@@ -427,4 +433,61 @@ setup_transient() {
     deploy_model "${OPENAI_MODEL_ID}" "${OPENAI_TASK_ID}" "OpenAI"
 }
 
+create_certificates() {
+    local HOST="${SSL_HOSTNAME:-localhost}"
+    local DAYS=10000
+
+    pushd config
+    local OLDMASK=$(umask -p)
+    umask 077
+
+    # 1. Make fake certificate authority (CA) certificate.  OpenSearch
+    # requires a root certificate to be specified.
+    openssl req -batch -x509 -newkey rsa:4096 -days "${DAYS}" \
+    -subj "/C=US/ST=California/O=Aryn.ai/CN=Fake CA" \
+    -extensions v3_ca -noenc -keyout cakey.pem -out cacert.pem 2> /dev/null
+    echo "Created CA certificate"
+
+    # 2a. Create certificate signing request (CSR) for the node certificate.
+    openssl req -batch -newkey rsa:4096 \
+    -subj "/C=US/ST=California/O=Aryn.ai/CN=${HOST}" \
+    -extensions v3_req -addext "basicConstraints=critical,CA:FALSE" \
+    -addext "subjectAltName=DNS:${HOST}" \
+    -noenc -keyout node-key.pem -out node-req.pem 2> /dev/null
+
+    # 2b. Use the fake CA to sign the node CSR, yielding a certificate.
+    openssl x509 -req -CA cacert.pem -CAkey cakey.pem \
+    -copy_extensions copy -days "${DAYS}" \
+    -in node-req.pem -out node-cert.pem 2> /dev/null
+    echo "Created node certificate"
+
+    # 3a. Create certificate signing request (CSR) for the admin certificate.
+    openssl req -batch -newkey rsa:4096 \
+    -subj "/C=US/ST=California/O=Aryn.ai/CN=Admin" \
+    -extensions v3_req -addext "basicConstraints=critical,CA:FALSE" \
+    -noenc -keyout admin-key.pem -out admin-req.pem 2> /dev/null
+
+    # 3b. Use the fake CA to sign the admin CSR, yielding a certificate.
+    openssl x509 -req -CA cacert.pem -CAkey cakey.pem \
+    -copy_extensions copy -days "${DAYS}" \
+    -in admin-req.pem -out admin-cert.pem 2> /dev/null
+    echo "Created admin certificate"
+
+    rm -f node-req.pem admin-req.pem
+
+    ${OLDMASK}
+    popd
+}
+
+setup_security() {
+    # Set up security plugin configuration
+    plugins/opensearch-security/tools/securityadmin.sh \
+    -cd config/opensearch-security -icl -nhnv -cacert config/cacert.pem \
+    -cert config/admin-cert.pem -key config/admin-key.pem
+
+    # Semaphore to signal completion
+    _curl -X PUT "${BASE_URL}/_cluster/settings" -o /dev/null --json \
+    '{"persistent":{"cluster":{"metadata":{"aryn_ssl_setup_complete":1}}}}'
```

Comment:
Does ssl support survive restarts or do you need to run the securityadmin command each time?
Where are you using the ssl_setup_complete semaphore?  I only see it here and not above.

For the semaphores should we have them be versioned, so use something like :20240206?  Then if we have to update the config we can tell what setup it was at (we could also put that off until the first time we have to do it).

Reply:
This survives restarts.  It would be pretty lousy security otherwise.

While the aryn_ssl_setup_complete semaphore isn't used in this script, it is used by stuff in pyscripts.  It's included here for consistency and semantic completeness.  After this milestone is when it's safe to register/deploy models.  If this is absent, it's a useful debugging clue.

Reply:
Why not use aryn_deploy_complete in pyscripts?
`# semaphore for debugging state if something goes wrong; can be used as a signal that it is ok to deploy models`

Reply:
The place in pyscripts that uses this is waiting for a different milestone in startup.  It's actually waiting for OpenSearch to be ready to register and deploy models.  I'll tweak the comment.